GraphConsoleAppV3/GraphRequests.cs

```diff
@@ -0,0 +1,1534 @@
+                                    }
+                                    registeredOwners = await registeredOwners.GetNextPageAsync();
+                                } while (registeredOwners != null);
+                            }
+                        }
+                    }
+                    devices = await devices.GetNextPageAsync();
+                } while (devices != null);
+            }
+
+            #endregion
+        }
+
+        private static async Task CreateOAuth2Permission(IActiveDirectoryClient client, ServicePrincipal newServicePrincpal)
+        {
+            #region Create a new consentable OAuth2 permission
+
+            //*********************************************************************************************
+            // Create new  oauth2 permission object
+            //*********************************************************************************************
+            if (newServicePrincpal.ObjectId != null)
+            {
+                OAuth2PermissionGrant permissionObject = new OAuth2PermissionGrant();
+                permissionObject.ConsentType = "AllPrincipals";
+                permissionObject.Scope = "user_impersonation";
+                permissionObject.StartTime = DateTime.Now;
+                permissionObject.ExpiryTime = (DateTime.Now).AddMonths(12);
+
+                // resourceId is objectId of the resource, in this case objectId of AzureAd (Graph API)
+                permissionObject.ResourceId = _graphAppObjectId;
+
+                //ClientId = objectId of servicePrincipal
+                permissionObject.ClientId = newServicePrincpal.ObjectId;
+
+                // add the oauth2 permission scope grant
+                try
+                {
+                    await client.Oauth2PermissionGrants.AddOAuth2PermissionGrantAsync(permissionObject);
+                    Console.WriteLine("New Permission object created: " + permissionObject.ObjectId);
+                }
+                catch (Exception e)
+                {
+                    Program.WriteError("\nError with Permission Creation: {0} {1}", e.Message,
+                        e.InnerException != null ? e.InnerException.Message : "");
+                }
+
+                // remove the oauth2 permission scope grant
+                try
+                {
+                    newServicePrincpal.Oauth2PermissionGrants.Remove(permissionObject);
+                    await newServicePrincpal.UpdateAsync();
+                    Console.WriteLine("Removed Permission object: " + permissionObject.ObjectId);
+                }
+                catch (Exception e)
+                {
+                    Program.WriteError("\nError with Permission Creation: {0} {1}", e.Message,
+                        e.InnerException != null ? e.InnerException.Message : "");
+                }
+
+                try
+                {
+                    await newServicePrincpal.DeleteAsync();
```

Comment:
>newServicePrincpal.DeleteAsync [](start = 26, length = 30)

you can pull this out in separate method and call it in finally block of UserMode()  along with other delete methods.